### PR TITLE
Closes #96 - Glycemic Status based on Lab data

### LINKED ISSUE #96: Feature Request: Glycemic Status based on Lab data
### Feature Idea

Glycemic Status includes "Normoglycemia", "Prediabetes" and "Diabetes" and can be derived as a new parameter in ADLB (or a separate ADaM dataset) as per [American Diabetes Association](https://doi.org/10.2337/dc25-S002) standards.

The derivation would involve input parameters such as HbA1c, Fasting Plasma Glucose and 2h Plasma Glucose during 75-g OGTT with an identification of two consecutive results above the threshold to consider a subject "Diabetic" at a certain time point.

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by yurovska**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral family codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [ ] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [ ] Update `NEWS.md` under the header `# admiral<ext> (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [ ] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by Siddhesh2097** [comment]
Minor comments from me @yurovska . Great work

**Comment by yurovska** [comment]
> Minor comments from me @yurovska . Great work

Thanks @Siddhesh2097

File: R/derive_adlb_params.R [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+#'   )
+#'   ```
+#'
+#'   where `xx.x` is a numeric threshold and names of the elements must match
+#'   `hba1c_code` and `fpg_code`.
+#'
+#' @permitted List of parameter-value pairs
+#'
+#' @details
+#' Implementation is based on ADA 2024.
```
Comment (AndersAskeland) [inline R/derive_adlb_params.R]: Should we try directly referencing the article? I.e. https://diabetesjournals.org/care/article/47/Supplement_1/S20/153954/2-Diagnosis-and-Classification-of-Diabetes. I am unsure how this is done in R packages, but can look it up.

File: R/admiralmetabolic-package.R
Comment (AndersAskeland) [inline R/admiralmetabolic-package.R]: We should think about whether we need to provide an example for the OGTT "criteria" as well ("2-h PG ≥200 mg/dL (≥11.1 mmol/L) during OGTT. The test should be performed as described by the WHO, using a glucose load containing the equivalent of 75 g anhydrous glucose dissolved in water.*")

File: R/derive_adlb_params.R [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+      id_cols = vars2chr(by_vars),
+      names_from = "PARAMCD",
+      names_glue = "{PARAMCD}.{.value}",
+      values_from = c("AVAL", "D_THRESHOLD")
+    ) %>%
+    # Identify cases where the criteria is met twice consecutively
+    derive_var_joined_exist_flag(
+      dataset = .,
+      dataset_add = .,
+      by_vars = get_admiral_option("subject_keys"),
```
Comment (AndersAskeland) [inline R/derive_adlb_params.R]: Would there be occasions where this would not work? I.e. there is no direct requirements to use admiral options to set the by vars.

**Review by AndersAskeland — COMMENTED** [review]
Looking good @yurovska ! 

I think we need a discussion if this is to trial specific. I.e. is this the way one always derive glycemic status. For instance, I wonder if this calculation is very study specific, where it differs between trials. From a diagnostic standpoint, there seems to be many different approaches to define this status.

File: R/admiralmetabolic-package.R
Comment (yurovska) [reply R/admiralmetabolic-package.R]: Waiting for #92

File: R/derive_adlb_params.R [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+                                 hba1c_code = "HBA1CHGB",
+                                 fpg_code = "GLUC",
+                                 get_unit_expr,
+                                 prediabetic_thresholds = rlang::list2(
+                                   !!hba1c_code := c("mmol/mol" = 39, "%" = 5.7),
+                                   !!fpg_code := c("mmol/L" = 5.6, "mg/dL" = 100)
+                                 ),
+                                 diabetic_thresholds = rlang::list2(
+                                   !!hba1c_code := c("mmol/mol" = 48, "%" = 6.5),
+                                   !!fpg_code := c("mmol/L" = 7, "mg/dL" = 126)
```
[the remark this answers is not in the document]
Comment (yurovska) [reply R/derive_adlb_params.R]: @AndersAskeland Bare `list()` does not support tidy evaluation (!! and :=).

@manciniedoardo We can write it as a `tibble` but what is the difference? I don't find it more clear.

```R
tibble::tibble(
  PARAMCD = c(!!hba1c_code, !!hba1c_code, !!fpg_code, !!fpg_code),
  UNIT = c("mmol/mol", "%", "mmol/L", "mg/dL"),
  THRESHOLD = c(39.0, 5.7, 5.6, 100.0)
)
```

File: R/derive_adlb_params.R [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+  diabetes_confirmed <- dataset %>%
+    filter(
+      PARAMCD %in% c(hba1c_code, fpg_code)
+    ) %>%
+    mutate(
+      # Map `diabetic_thresholds` into a variable
+      D_THRESHOLD = purrr::map2_dbl(
+        PARAMCD, !!get_unit_expr,
+        ~ purrr::pluck(diabetic_thresholds, .x, .y) %||%
+          purrr::pluck(diabetic_thresholds, .x, .default = NA_real_)
```
[the remark this answers is not in the document]
Comment (yurovska) [reply R/derive_adlb_params.R]: It's not the same as `PRED_THRESHOLD` above as it uses different thresholds.

File: R/derive_adlb_params.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (yurovska) [reply R/derive_adlb_params.R]: @manciniedoardo You're overcomplicating things. The only reason the thresholds are there in the function arguments is to avoid hardcoding. The chance that users would ever need to deviate from ADA criteria and pass those arguments in their function call is practically equal to 0%. Tidy evaluation is needed here to spare the user from having to pass them to the function in cases where there are non-default parameter codes, so that when the user changes `hba1c_code` and/or `fpg_code` to non-default ones, the `prediabetic_thresholds` and `diabetic_thresholds` will still work with no need to create any user-defined metadata.

Even if in some parallel universe there is a user who needs to deviate from the standard ADA criteria, they can simply provide a bare list without using tidy evaluation, and they won't even have to provide more than one unit. This is already described clearly in the documentation.

image
